### PR TITLE
Render modules synopsis for `{!modules:...}`

### PR DESCRIPTION
**Pull request by Julow**
Fixes https://github.com/ocaml/odoc/issues/297

Resolve and render the synopsis of modules referenced with the `{!modules:A B C}` markup.

The first sentence of the first paragraph of the modules' documentation is taken. Headings, tags and other elements that are not paragraphs and lists are skipped, I think that's what ocamldoc does: https://caml.inria.fr/pub/docs/manual-ocaml/ocamldoc.html#sss:ocamldoc-preamble

### CODE REVIEW

**Comment by Drup** [comment]
Looks good on my side :+1:

@dbuenzli An opinion on the resulting HTML ?

File: src/model/comment.ml [hunk 1]
```diff
@@ -29,11 +29,20 @@ type inline_element =
   | `Reference of Reference.t * link_content
   | `Link of string * link_content ]
 
+type paragraph = inline_element with_location list
+
+type module_reference = {
+  module_reference : Reference.Module.t;
+  module_synopsis : paragraph option;
+}
```
Comment (Julow) [inline src/model/comment.ml:37]: This module is used for two things: the output of the parser and the representation of comments inside Odoc. Here I intent to change the latter but it's a breaking change for users of the parser (eg. OCamlformat and Mdx).

A fix to that would be to duplicate the previous version of this module into the parser library and adding a conversion step.
We ought to extract the parser into its own package, that would be a breaking change too and would require that extra step anyway.

**Comment by Julow** [comment]
I added the `synopsis` class too.

File: src/model/comment.ml (hunk 1 above)
Comment (jonludlam) [reply src/model/comment.ml:37]: Don't mdx and ocamlformat use `Ast.docs` rather than `Comment.docs`?

**Comment by dbuenzli** [comment]
According to what I see this uses definition lists, please don't use definitions list, they [needlessly hard to style](https://github.com/ocaml/odoc/issues/329#issuecomment-781271998). Use unordered lists as they were before. 

Also there's no need to put the synopsis in its own paragraph as it should be only one paragraph; just `span` it after the module name (don't forget to add a space between the module name and the synopsis `span`).

Basically what you want is the same markup as [odig's lists of packages](https://b0-system.github.io/odig/doc/). *Mutatis mutandis* that would be: 

```html
<ul class="modules">
  <li id="module-M">
    <a href="#module-M" class="anchor"></a>
    <a href="M/index.html"><code>M</code></a>
    <span class="synopsis">M's synopsis</span>
  </li>
</ul>
```

**Comment by Drup** [comment]
If we should never use definition lists in HTML, I suggest we change the HTML renderer rather. The structure should be preserved internally (for the Latex output, for instance).

**Comment by dbuenzli** [comment]
I guess so, I was actually confused not to see anything in the HTML renderer so I eventually went to the samples (I'm not up-to-date with the intermediate representation business).

**Comment by dbuenzli** [comment]
> The first sentence of the first paragraph of the modules' documentation is taken.

Detecting the first sentence is very unreliable. I think we had mostly agreed with @lpw25 that `odoc` would use the first paragraph. Here's a [discussion](https://github.com/ocaml/ocaml/pull/784) about it.

**Comment by gasche** [comment]
Thanks @Julow for the implementation work!

I am fine with the "first paragraph" specification, I agree it is easier and it also sounds more convenient for users anyway -- what if I want several sentences in my synopsis? In practice I see people breaking the paragraph after the synopsis, so I think that most libraries would not notice any difference compared to ocamldoc's behavior.

**Comment by jonludlam** [comment]
Thanks for the reference, @dbuenzli - I'm _much_ happier to use the first paragraph than try to figure out where the first sentence ends!

We should probably add this difference to our `interfaces` doc so it's clear we're behaving a little differently than ocamldoc here.

File: src/model/comment.ml (hunk 1 above)
Comment (Julow) [reply src/model/comment.ml:37]: You are right ! I invented the problem.

**Comment by Julow** [comment]
I've changed to "first paragraph".
